docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+* Use a different enum for non-error cases with multiple return types (which can't use Option).
+
+# Advanced Features
+
+## Operator Overloading
+
+### No clever operators :: required
+
+Other than the comparison operators defined by [**Eq**](https://doc.rust-lang.org/std/cmp/trait.Eq.html), [**Ord**](https://doc.rust-lang.org/std/cmp/trait.Ord.html) etc. there is no obvious benefit to overloading other operators (e.g. overloading `*` to do something clever via the [Mul](https://doc.rust-lang.org/std/ops/trait.Mul.html) trait). This could be relaxed if standard Rust idioms for using particular operator overloads exists, and is well understood in the Rust community.
+
+See also [Operators and Symbols](https://doc.rust-lang.org/book/appendix-02-operators.html).
+
+## Binding Traits to Inbuilt Types
+
+Rust has neat ways to provide users with simple type conversion. Imagine if you have a function taking (reference to) a semantic type:
+
+```
+fn use_locale_id(id: &LocaleId) { … }
+```
+
+But the caller could just write:
+
+```
+use_locale_id(&"en_GB".into());
+```
+
+Rust's lets you bind traits to existing system type (e.g. `str`) for use within a module. And importantly, it lets you expose a series to trait bindings that other people can opt into if they want.
+By implementing the generic [`From<&str>`](https://doc.rust-lang.org/std/convert/trait.From.html) trait on `LocaleId` to convert from a string to a locale ID, we also get the [`Into<Foo>`](https://doc.rust-lang.org/std/convert/trait.Into.html) trait implied on `&str` for free.
+
+```
+impl From<&str> for LocaleId {
+  fn from(s: &str) -> LocaleId {
+    // This is just the normal object creation function on LocaleId.
+    LocaleId::from(s)
+  }
+}
+```
+
+Note that there's also the [FromStr](https://doc.rust-lang.org/beta/std/str/trait.FromStr.html) trait for things which are explicitly parseable from strings.
+
+However you can do more than just conversion types, and the [Unicode Segmentation](https://crates.io/crates/unicode-segmentation) crate binds a trait with many functions onto `str` to allow users to write things like:
+
+```
+use unicode_segmentation::UnicodeSegmentation;
+
+// A vector of individual graphemes (true => extended).
+let graphemes = "a\r\nb🇷🇺🇸🇹".graphemes(true).collect::<Vec<&str>>();
+
+// Print each word (according to Unicode).
+let s = "The quick (\"brown\") fox can't jump 32.3 feet, right?";
+for w in s.unicode_words() {
+  println!("{:?}", w);
+}
+```
+
+Thus we could provide one or more ICU4X traits bound to things like `str` to provide a low friction way to access the libraries (obvious questions like naming notwithstanding).
+
+**Open Question**: Is this too much "cleverness" ? Personally I like this a lot because it's safe and opt-in for users (there can always be plain functions to just call, and bindings like this could be exported in a separate module so users aren't forced to adopt them).
```

Comment:
conversion functions are pretty good and normal. extension traits are also a normal pattern. Both are fine.

Reply:
I've just removed this question - it seems people are fine with it.